### PR TITLE
nixos/switch-to-configuration: Major script cleanup

### PR DESCRIPTION
**Pull request by dasJ**
###### Description of changes

Shorter reviewer list because I don't know more people with Perl skills.
This should not change any logic at all.

It's useful to make the code more verbose and longer to make it more readable since most people don't know Perl very well.
This is to make the logic more accessible to more people who are not used to stuff like postfix ifs and `unless` blocks.

Some of the commits are also to improve consistency in the code and variable naming, which is just nice to have and to add comments.
Also, Perlcritic is followed on the 1 level quite a lot but a lot of checks that were too opinionated were disabled.

The entire point of this PR is to split cleanups from logic changes to not bury any relevant logic changes in the cleanup.

Best reviewed by commit ;)

@GrahamcOfBorg test switchTest

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/system/activation/switch-to-configuration.pl [hunk 1]
```diff
@@ -580,31 +586,31 @@ sub path_to_unit_name {
 # automatically by starting local-fs.target.  FIXME: might be nicer if
 # we generated units for all mounts; then we could unify this with the
 # unit checking code above.
-my ($prevFss, $prevSwaps) = parse_fstab("/etc/fstab");
-my ($newFss, $newSwaps) = parse_fstab("$out/etc/fstab");
-foreach my $mountPoint (keys(%$prevFss)) {
-    my $prev = $prevFss->{$mountPoint};
-    my $new = $newFss->{$mountPoint};
-    my $unit = path_to_unit_name($mountPoint);
+my ($cur_fss, $prev_swaps) = parse_fstab("/etc/fstab");
```
Comment (ajs124) [inline nixos/modules/system/activation/switch-to-configuration.pl:597]: why rename `prevFss` to `cur_fss` but `prevSwaps` to `prev_swaps`?

File: nixos/modules/system/activation/switch-to-configuration.pl [hunk 2]
```diff
@@ -470,13 +471,13 @@ sub handleModifiedUnit {
...
-    split('\n', read_file($startListFile, err_mode => 'quiet') // "");
+    split("\n", read_file($startListFile, err_mode => "quiet") // "");
 
 %unitsToRestart = map { $_ => 1 }
-    split('\n', read_file($restartListFile, err_mode => 'quiet') // "");
+    split("\n", read_file($restartListFile, err_mode => "quiet") // "");
 
 %unitsToReload = map { $_ => 1 }
-    split('\n', read_file($reloadListFile, err_mode => 'quiet') // "");
+    split("\n", read_file($reloadListFile, err_mode => "quiet") // "");
```
Comment (ncfavier) [inline nixos/modules/system/activation/switch-to-configuration.pl]: Careful, `"\n" != '\n'`.

File: nixos/modules/system/activation/switch-to-configuration.pl [hunk 3]
```diff
@@ -95,7 +100,7 @@
 my $new_init_interface_version = read_file("$out/init-interface-version");
 
 if ($new_init_interface_version ne $cur_init_interface_version) {
-    print STDERR <<EOF;
+    print STDERR <<'EOF';
 Warning: the new NixOS configuration has an ‘init’ that is
 incompatible with the current configuration.  The new configuration
 won\'t take effect until you reboot the system.
```
Comment (ncfavier) [inline nixos/modules/system/activation/switch-to-configuration.pl:106]: Can unescape the `\'` now.

File: nixos/modules/system/activation/switch-to-configuration.pl [hunk 4]
```diff
@@ -385,21 +392,21 @@ sub compare_units {
...
         # properties (resource limits and inotify watches)
         # seem to get applied on daemon-reload.
-    } elsif ($unit =~ /\.mount$/) {
+    } elsif ($unit =~ /\.mount$/msx) {
         # Reload the changed mount unit to force a remount.
         # FIXME: only reload when Options= changed, restart otherwise
         $units_to_reload->{$unit} = 1;
         record_unit($reload_list_file, $unit);
-    } elsif ($unit =~ /\.socket$/) {
+    } elsif ($unit =~ /\.socket$/msx) {
```
Comment (ncfavier) [inline nixos/modules/system/activation/switch-to-configuration.pl:409]: Not a fan of the `/msx` everywhere, especially on clearly single-line strings. Why is that required?

File: nixos/modules/system/activation/switch-to-configuration.pl [hunk 5]
```diff
@@ -873,7 +881,9 @@ sub filter_units {
 
     if ($state->{substate} eq "auto-restart") {
         # A unit in auto-restart substate is a failure *if* it previously failed to start
-        my $main_status = `$new_systemd/bin/systemctl show --value --property=ExecMainStatus '$unit'`;
+        open(my $main_status_fd, "-|", "$new_systemd/bin/systemctl", "show", "--value", "--property=ExecMainStatus", $unit) || die("Unable to call 'systemctl show'");
+        my $main_status = do { local $/ = undef; <$main_status_fd> };
+        close($main_status_fd) || die("Unable to close 'systemctl show' fd");
```
Comment (ncfavier) [inline nixos/modules/system/activation/switch-to-configuration.pl:886]: Why not just `my $main_status = system("$new_systemd/bin/systemctl", "show", ...);` ?

**Review by ncfavier — CHANGES_REQUESTED** [review]

**Comment by ncfavier** [comment]
Don't forget to snake-case the newly added `installBootLoader` variable

File: nixos/modules/system/activation/switch-to-configuration.pl (hunk 2 above)
Comment (dasJ) [reply nixos/modules/system/activation/switch-to-configuration.pl]: A later commit does exactly that ;)

Comment (ncfavier) [reply nixos/modules/system/activation/switch-to-configuration.pl]: Heh, I don't know how I missed it

Comment (dasJ) [reply nixos/modules/system/activation/switch-to-configuration.pl]: Don't worry about it, I was confused as well because I couldn't find `'\n'` in the file :D

File: nixos/modules/system/activation/switch-to-configuration.pl (hunk 4 above)
Comment (ncfavier) [reply nixos/modules/system/activation/switch-to-configuration.pl:409]: OK, it's fine

**Review by ncfavier — APPROVED** [review]
@ofborg test switchTest

**Comment by ncfavier** [comment]
I said
@ofborg test switchTest

File: nixos/modules/system/activation/switch-to-configuration.pl [hunk 6]
```diff
@@ -315,11 +355,11 @@ sub compare_units {
                 return 1;
             }
         }
-        # A key was introduced that was missing in the old unit
+        # A key was introduced that was missing in the previos unit
```
Comment (cheriimoya) [inline nixos/modules/system/activation/switch-to-configuration.pl:358]:
```suggestion
        # A key was introduced that was missing in the previous unit
```

File: nixos/modules/system/activation/switch-to-configuration.pl [hunk 7]
```diff
@@ -332,13 +372,13 @@ sub compare_units {
             }
         };
     }
-    # A section was introduced that was missing in the old unit
+    # A section was introduced that was missing in the previos unit
```
Comment (cheriimoya) [inline nixos/modules/system/activation/switch-to-configuration.pl:375]:
```suggestion
    # A section was introduced that was missing in the previous unit
```

File: nixos/modules/system/activation/switch-to-configuration.pl [hunk 8]
```diff
@@ -350,76 +390,78 @@ sub compare_units {
     return $ret;
 }
 
-sub handleModifiedUnit {
-    my ($unit, $baseName, $newUnitFile, $newUnitInfo, $activePrev, $unitsToStop, $unitsToStart, $unitsToReload, $unitsToRestart, $unitsToSkip) = @_;
+# Called when a unit exists in both the old systemd and the new system and the units
+# differ. This takes care of what units are to be stopped, restarted, reloaded, started, and skipped.
```
Comment (cheriimoya) [inline nixos/modules/system/activation/switch-to-configuration.pl:394]: I don't really understand the meaning of this. Does it "take care" in terms of taking action? If so:

```suggestion
# differ. This takes of stopping, restarting, reloading, starting, and skipping units.
```

File: nixos/modules/system/activation/switch-to-configuration.pl [hunk 9]
```diff
@@ -592,97 +639,109 @@ sub pathToUnitName {
...
-    $restartSystemd = 1;
+my $restart_systemd = $cur_pid1_path ne $new_pid1_path;
+if ($cur_systemd_system_config ne $new_systemd_system_config) {
+    $restart_systemd = 1;
 }
 
-
-sub filterUnits {
+# Takes an array of unit names and returns an array with the same elements,
+# but all units that are also in the global variable `units_to_filter` removed.
```
Comment (cheriimoya) [inline nixos/modules/system/activation/switch-to-configuration.pl:653]:
```suggestion
# except all units that are also in the global variable `units_to_filter`.
```

**Review by cheriimoya — CHANGES_REQUESTED** [review]
